File: algorithms/cpp/329.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
private:
    int dfs(int i, int j, vector<vector<int>> &matrix, vector<vector<int>> &dp) {
        if (dp[i][j] > 1)
            return dp[i][j];
        int n = matrix.size(), m = matrix[0].size();
        if (i-1 >= 0 && matrix[i-1][j] < matrix[i][j])
            dp[i][j] = max(dp[i][j], dfs(i-1, j, matrix, dp)+1);
        if (j-1 >= 0 && matrix[i][j-1] < matrix[i][j])
            dp[i][j] = max(dp[i][j], dfs(i, j-1, matrix, dp)+1);
        if (i+1 < n && matrix[i+1][j] < matrix[i][j])
            dp[i][j] = max(dp[i][j], dfs(i+1, j, matrix, dp)+1);
        if (j+1 < m && matrix[i][j+1] < matrix[i][j])
            dp[i][j] = max(dp[i][j], dfs(i, j+1, matrix, dp)+1);
        return dp[i][j];
    }
public:
    int longestIncreasingPath(vector<vector<int>> &matrix) {
        if (matrix.size() == 0)
            return 0;
        int n = matrix.size(), m = matrix[0].size();
        vector<vector<int>> dp(n, vector<int>(m, 1));
        int result = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                result = max(result, dfs(i, j, matrix, dp));
            }
        }
        return result;
    }
};

int main() {
    return 0;
}

File: algorithms/python/329.py
class Solution:
    def longestIncreasingPath(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        if not len(matrix) or not len(matrix[0]):
            return 0

        m, n = len(matrix), len(matrix[0])
        pq = sorted([(matrix[i][j], i, j) for i in range(m) for j in range(n)])

        rls = [[1 for i in range(n)] for j in range(m)]
        for v, x, y in pq:
            if x-1 >= 0 and matrix[x-1][y] > v:
                rls[x-1][y] = max(rls[x-1][y], rls[x][y]+1)
            if x+1 < m and matrix[x+1][y] > v:
                rls[x+1][y] = max(rls[x+1][y], rls[x][y]+1)
            if y-1 >= 0 and matrix[x][y-1] > v:
                rls[x][y-1] = max(rls[x][y-1], rls[x][y]+1)
            if y+1 < n and matrix[x][y+1] > v:
                rls[x][y+1] = max(rls[x][y+1], rls[x][y]+1)

        return max(map(max, rls))
